biapol_utilities/label/_matching_algorithms.py

```diff
@@ -0,0 +1,259 @@
+        corresponds to the "similarity" between label i in label_image_x and
+        label j in label_image_y.
+
+    Returns
+    -------
+    ND-array
+        Relabelled version of input_image_y. Unmatched labels are appended to
+        the set of matched labels and thus, the total number of labels may
+        increase.
+
+    """
+    # Remove background-label from similarity matrix
+    similarity_matrix = similarity_matrix[1:, 1:]
+
+    # Suppress non-maximal entries
+    similarity_matrix[similarity_matrix < similarity_matrix.max(axis=0)] = 0.0
+
+    mmax = label_image_x.max()
+
+    if similarity_matrix.size > 0:
+        # Pick value with highest IoU value
+        istitch = similarity_matrix.argmax(axis=1) + 1
+
+        # Find unpaired labels
+        ino = np.nonzero(similarity_matrix.max(axis=1) == 0.0)[0]
+
+        # append unmatched labels and background to lookup table
+        istitch[ino] = np.arange(mmax+1, mmax+len(ino)+1, 1, int)
+        mmax += len(ino)
+        istitch = np.append(np.array(0), istitch)
+
+        return istitch[label_image_y]
+
+    else:
+        raise ValueError('Similarity matrix dimension was expected to have '
+                         'size MxM but was found to have size '
+                         f'{similarity_matrix.size}. Check calculation of'
+                         'similarity matrix for errors.')
+
+
+def gale_shapley(label_image_x, label_image_y, similarity_matrix):
+    """
+    Gale-Shapley stable-marriage algorithm for label-matching.
+
+    Implementation of Gale-Shapley's solution of the stable-marriage problem
+    [0, 1] to match labels from two label images (`label_image_x` and
+    `label_image_y`) based on a mutual set of preferences that are passed as
+    `similarity_matrix`. Unmatchable labels in `label_image_x` maintain their
+    previous labels whereas unmatchable labels in `label_image_y` may be
+    assigned to new labels, thus potentially increasing the total number of
+    labels.
+
+    Parameters
+    ----------
+    label_image_x : ND-array, int
+        label image that serves as reference label image
+    label_image_y : ND-array, int
+        Label image, the labels of which are assigned to `label_image_x`
+    similarity_matrix : NxN, float
+        similarity matrix where `similarity_matrix[i, j]` corresponds to the
+        similarity/preference to match the i-th and j-th entry from the set of
+        labels `set(label_image_x, label_image_y)` with each other.
+
+    Returns
+    -------
+    ND-array, int
+        Copy of label_image_y with labels assigned to labels_image_x.
+
+    References
+    ----------
+    .. [0] https://en.wikipedia.org/wiki/Gale%E2%80%93Shapley_algorithm
+    .. [1] https://doi.org/10.1080/00029890.1962.11989827
+
+    """
+    # Get set of input labels from both datasets and combined labels
+    list_of_men = np.unique(np.append(0, label_image_x.ravel()))
+    list_of_women = np.unique(np.append(0, label_image_y.ravel()))
```

Comment:
Please use "x" and "y" instead of men and women. It's confusing like it is now.

Reply:
Changed in 6fd5335ffdb83b40cd949a59118ad5d4837cbf54